application/src/main/java/org/thingsboard/server/service/alarm/rule/DefaultTbAlarmRuleStateService.java

```diff
@@ -0,0 +1,517 @@
+import org.thingsboard.server.common.data.id.HasId;
+import org.thingsboard.server.common.data.id.RuleChainId;
+import org.thingsboard.server.common.data.id.RuleNodeId;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.page.PageDataIterable;
+import org.thingsboard.server.common.data.plugin.ComponentLifecycleEvent;
+import org.thingsboard.server.common.data.rule.RuleNode;
+import org.thingsboard.server.common.data.util.CollectionsUtil;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.plugin.ComponentLifecycleMsg;
+import org.thingsboard.server.common.msg.queue.ServiceType;
+import org.thingsboard.server.common.msg.queue.TopicPartitionInfo;
+import org.thingsboard.server.dao.alarm.rule.AlarmRuleService;
+import org.thingsboard.server.gen.transport.TransportProtos;
+import org.thingsboard.server.queue.discovery.event.PartitionChangeEvent;
+import org.thingsboard.server.queue.util.TbRuleEngineComponent;
+import org.thingsboard.server.service.alarm.rule.state.PersistedEntityState;
+import org.thingsboard.server.service.alarm.rule.store.AlarmRuleEntityStateStore;
+import org.thingsboard.server.service.partition.AbstractPartitionBasedService;
+import org.thingsboard.server.service.profile.TbAssetProfileCache;
+import org.thingsboard.server.service.profile.TbDeviceProfileCache;
+
+import javax.annotation.PostConstruct;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.stream.Collectors;
+
+import static org.thingsboard.server.common.data.DataConstants.INTERNAL_QUEUE_NAME;
+import static org.thingsboard.server.common.data.plugin.ComponentLifecycleEvent.DELETED;
+
+@Slf4j
+@Service
+@TbRuleEngineComponent
+@RequiredArgsConstructor
+public class DefaultTbAlarmRuleStateService extends AbstractPartitionBasedService<EntityId> implements TbAlarmRuleStateService {
+
+    private final TbAlarmRuleContext ctx;
+    private final AlarmRuleService alarmRuleService;
+    private final AlarmRuleEntityStateStore stateStore;
+    //    private final RelationService relationService;
+    private final TbDeviceProfileCache deviceProfileCache;
+    private final TbAssetProfileCache assetProfileCache;
+
+    private final Map<EntityId, EntityState> entityStates = new ConcurrentHashMap<>();
+    private final Map<AlarmRuleId, Set<EntityId>> myEntityStateIdsPerAlarmRule = new ConcurrentHashMap<>();
+    private final Map<TenantId, Set<EntityId>> myEntityStateIds = new ConcurrentHashMap<>();
+    private Cache<TenantId, ConcurrentHashMap<AlarmRuleId, AlarmRule>> rulesCache;
```

Comment:
Probably would be better not to store map of maps, so consider creating a composite key of tenantId and alarmRuleId and store everything in the Caffeine cache.

Reply:
`Map<TenantId, Cache<AlarmRuleId, AlarmRule>> caches;`